rocketpy/motors/Tank.py

```diff
@@ -0,0 +1,524 @@
+        self.liquid_height = Function(
+            liquid_height, "Time", "Volume", "linear", "constant"
+        )
+
+    @funcify_method("Time (s)", "mass (kg)")
+    def mass(self):
+        return self.liquidMass + self.gasMass
+
+    @funcify_method("Time (s)", "mass flow rate (kg/s)")
+    def netMassFlowRate(self):
+        return self.mass.derivativeFunction()
+
+    @funcify_method("Time (s)", "volume (m³)")
+    def liquidVolume(self):
+        return self.geometry.volume.compose(self.liquidHeight)
+
+    @funcify_method("Time (s)", "volume (m³)")
+    def gasVolume(self):
+        return self.geometry.total_volume - self.liquidVolume
+
+    @funcify_method("Time (s)", "height (m)")
+    def liquidHeight(self):
+        return self.liquid_height
+
+    @funcify_method("Time (s)", "mass (kg)")
+    @except_negative
+    def gasMass(self, t):
+        return self.gasVolume(t) * self.gas.density
+
+    @funcify_method("Time (s)", "mass (kg)")
+    @except_negative
+    def liquidMass(self, t):
+        return self.liquidVolume(t) * self.liquid.density
+
+    @funcify_method("Time (s)", "height (m)")
+    def gasHeight(self, t):
+        return self.geometry.top
+
+
+class MassBasedTank(Tank):
+    def __init__(
+        self,
+        name,
+        geometry,
+        liquid,
+        gas,
+        liquid_mass,
+        gas_mass,
+    ):
+        super().__init__(name, geometry, gas, liquid)
+        self.liquid_mass = Function(liquid_mass, "Time", "Mass", "linear", "constant")
+        self.gas_mass = Function(gas_mass, "Time", "Mass", "linear", "constant")
+
+    @funcify_method("Time (s)", "mass (kg)")
+    def mass(self):
+        return self.liquidMass + self.gasMass
+
+    @funcify_method("Time (s)", "mass flow rate (kg/s)")
+    def netMassFlowRate(self):
+        return self.mass.derivativeFunction()
+
+    @funcify_method("Time (s)", "mass (kg)")
+    @except_negative
+    def liquidMass(self, t):
+        return self.liquid_mass
+
+    @funcify_method("Time (s)", "mass (kg)")
+    @except_negative
+    def gasMass(self, t):
+        return self.gas_mass
+
+    @funcify_method("Time (s)", "volume (m³)")
+    def gasVolume(self):
+        return self.gasMass / self.gas.density
+
+    @funcify_method("Time (s)", "volume (m³)")
+    def liquidVolume(self):
+        return self.liquidMass / self.liquid.density
+
+    @funcify_method("Time (s)", "height (m)")
+    def liquidHeight(self):
+        return self.geometry.inverse_volume.compose(self.liquidVolume)
+
+    @funcify_method("Time (s)", "height (m)")
+    def gasHeight(self, t):
+        fluid_volume = self.gasVolume + self.liquidVolume
+        gasHeight = self.geometry.inverse_volume(fluid_volume(t))
+        if gasHeight <= self.geometry.top:
+            return gasHeight
+        else:
+            raise ValueError(
+                f"The tank {self.name}, is overfilled"
+                f"with gas height {gasHeight} at time {t}"
+            )
```

Comment:
This overfilling check needs to be done in `liquidHeight` too.

Reply:
@phmbressan , any comment?

Reply:
This is a good catch. To be clear, my earlier reasoning for not doing this check was that, in code, `gasHeight >= liquidHeight`.

Therefore, if the liquidHeight verfilled, the gasHeight exception would trigger. However, this may not happen, if, for some reason, the property `liquidHeight` is called alone.

This is indeed a mistake.

Reply:
Solved here: [3cb8e8d](https://github.com/RocketPy-Team/RocketPy/pull/233/commits/3cb8e8d81b49c0e1925ab80f0a4d6159952b7a90)